File: nlp_semantics/semantics.py
"""Semantic analysis engine."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Iterable

from nlp_semantics import lexicon

WORD_RE = re.compile(r"[A-Za-z']+|[0-9]+|[^\sA-Za-z0-9]")


@dataclass(frozen=True)
class Token:
    text: str
    lemma: str
    index: int


def tokenize(sentence: str) -> list[Token]:
    matches = WORD_RE.findall(sentence)
    tokens = []
    for index, match in enumerate(matches):
        lemma = match.lower()
        lemma = lemma.replace("'", "")
        tokens.append(Token(text=match, lemma=lemma, index=index))
    return tokens


def _is_content_word(lemma: str) -> bool:
    return lemma.isalpha() and lemma not in lexicon.DETERMINERS and lemma not in lexicon.PRONOUNS


def _guess_pos(lemma: str) -> str:
    if lemma in lexicon.COMMON_VERBS:
        return "verb"
    if lemma.endswith("s") and lemma[:-1] in lexicon.COMMON_VERBS:
        return "verb"
    if lemma.endswith(lexicon.VERB_SUFFIXES):
        return "verb"
    if lemma.endswith(lexicon.ADVERB_SUFFIXES):
        return "adverb"
    if lemma.endswith(lexicon.ADJECTIVE_SUFFIXES):
        return "adjective"
    if lemma in lexicon.DETERMINERS:
        return "determiner"
    if lemma in lexicon.PRONOUNS:
        return "pronoun"
    if lemma in lexicon.QUESTION_WORDS:
        return "question_word"
    if lemma.isdigit():
        return "number"
    if not lemma.isalpha():
        return "punctuation"
    return "noun"


def _detect_sentence_type(tokens: Iterable[Token]) -> str:
    if any(token.text == "?" for token in tokens):
        return "interrogative"
    if any(token.text == "!" for token in tokens):
        return "exclamatory"
    return "declarative"


def _estimate_tense(tokens: list[Token]) -> str:
    lemmas = [token.lemma for token in tokens]
    if any(token.text.lower().endswith("ed") for token in tokens):
        return "past"
    if any(lemma == "will" for lemma in lemmas):
        return "future"
    if any(lemma in ("was", "were", "had") for lemma in lemmas):
        return "past"
    return "present"


def _pick_main_verb(tokens: list[Token]) -> Token | None:
    for token in tokens:
        if _guess_pos(token.lemma) == "verb":
            return token
    return None


def _guess_subject(tokens: list[Token], verb_index: int | None) -> str | None:
    if verb_index is None:
        return None
    candidates = [token for token in tokens[:verb_index] if _guess_pos(token.lemma) in {"noun", "pronoun"}]
    if candidates:
        return " ".join(token.text for token in candidates[-2:])
    return None


def _guess_object(tokens: list[Token], verb_index: int | None) -> str | None:
    if verb_index is None:
        return None
    candidates = [token for token in tokens[verb_index + 1 :] if _guess_pos(token.lemma) in {"noun", "pronoun"}]
    if candidates:
        return " ".join(token.text for token in candidates[:2])
    return None


def _sentiment_score(tokens: list[Token]) -> int:
    score = 0
    for token in tokens:
        lemma = token.lemma
        if lemma in lexicon.SENTIMENT_POSITIVE:
            score += 1
        if lemma in lexicon.SENTIMENT_NEGATIVE:
            score -= 1
    return score


def _sentiment_label(score: int) -> str:
    if score > 0:
        return "positive"
    if score < 0:
        return "negative"
    return "neutral"


def _collect_topics(tokens: list[Token]) -> list[str]:
    topics = []
    for token in tokens:
        if _is_content_word(token.lemma) and token.lemma.isalpha():
            topics.append(token.lemma)
    return list(dict.fromkeys(topics))


def _modality(tokens: list[Token]) -> list[str]:
    modalities = []
    for token in tokens:
        if token.lemma in lexicon.MODALITY:
            modalities.append(lexicon.MODALITY[token.lemma])
    return modalities


def _polarity(tokens: list[Token]) -> str:
    if any(token.lemma in lexicon.NEGATIONS for token in tokens):
        return "negative"
    return "affirmative"


def _temporal_cues(tokens: list[Token]) -> list[str]:
    cues = []
    for token in tokens:
        if token.lemma in lexicon.TEMPORAL_CUES:
            cues.append(token.lemma)
    return cues


def _quantifiers(tokens: list[Token]) -> list[str]:
    quantifiers = []
    for token in tokens:
        if token.lemma in lexicon.QUANTIFIERS:
            quantifiers.append(token.lemma)
    return quantifiers


def _word_influence(tokens: list[Token]) -> list[dict[str, object]]:
    influences = []
    for token in tokens:
        lemma = token.lemma
        pos = _guess_pos(lemma)
        influence = "background"
        weight = 0.2
        if _is_content_word(lemma):
            influence = "content"
            weight = 1.0
        if lemma in lexicon.NEGATIONS:
            influence = "negation"
            weight = 1.4
        if lemma in lexicon.INTENSIFIERS:
            influence = "intensifier"
            weight = 1.2
        if lemma in lexicon.DIMINISHERS:
            influence = "diminisher"
            weight = 0.6
        if lemma in lexicon.QUANTIFIERS:
            influence = "quantifier"
            weight = 1.1
        if lemma in lexicon.MODALITY:
            influence = "modality"
            weight = 1.1
        influences.append(
            {
                "token": token.text,
                "lemma": lemma,
                "index": token.index,
                "pos_guess": pos,
                "influence_type": influence,
                "weight": weight,
            }
        )
    return influences


def analyze_sentence(sentence: str) -> dict[str, object]:
    tokens = tokenize(sentence)
    main_verb = _pick_main_verb(tokens)
    verb_index = main_verb.index if main_verb else None
    sentiment_score = _sentiment_score(tokens)
    analysis = {
        "sentence": sentence,
        "sentence_type": _detect_sentence_type(tokens),
        "polarity": _polarity(tokens),
        "modality": _modality(tokens),
        "tense": _estimate_tense(tokens),
        "temporal_cues": _temporal_cues(tokens),
        "quantifiers": _quantifiers(tokens),
        "main_predicate": main_verb.text if main_verb else None,
        "subject": _guess_subject(tokens, verb_index),
        "object": _guess_object(tokens, verb_index),
        "topics": _collect_topics(tokens),
        "sentiment": {
            "score": sentiment_score,
            "label": _sentiment_label(sentiment_score),
        },
        "word_influence": _word_influence(tokens),
    }
    return analysis


def format_analysis(analysis: dict[str, object]) -> str:
    lines = ["Semantic Analysis"]
    lines.append("=" * 18)
    lines.append(f"Sentence: {analysis['sentence']}")
    lines.append(f"Type: {analysis['sentence_type']}")
    lines.append(f"Polarity: {analysis['polarity']}")
    lines.append(f"Modality: {', '.join(analysis['modality']) or 'none'}")
    lines.append(f"Tense: {analysis['tense']}")
    lines.append(f"Temporal cues: {', '.join(analysis['temporal_cues']) or 'none'}")
    lines.append(f"Quantifiers: {', '.join(analysis['quantifiers']) or 'none'}")
    lines.append("Predicate/Arguments:")
    lines.append(f"  Main verb: {analysis['main_predicate'] or 'unknown'}")
    lines.append(f"  Subject: {analysis['subject'] or 'unknown'}")
    lines.append(f"  Object: {analysis['object'] or 'unknown'}")
    lines.append(f"Topics: {', '.join(analysis['topics']) or 'none'}")
    sentiment = analysis["sentiment"]
    lines.append(f"Sentiment: {sentiment['label']} (score {sentiment['score']})")
    lines.append("\nWord influence:")
    for item in analysis["word_influence"]:
        lines.append(
            "  - {token} ({pos_guess}): {influence_type} (weight {weight})".format(**item)
        )
    return "\n".join(lines)

File: webapp/src/semantics.js
const NEGATIONS = new Set([
  "not",
  "no",
  "never",
  "none",
  "nobody",
  "nothing",
  "neither",
  "nowhere",
  "hardly",
  "scarcely",
  "barely"
]);

const INTENSIFIERS = new Set([
  "very",
  "extremely",
  "really",
  "so",
  "too",
  "quite",
  "highly",
  "deeply",
  "utterly"
]);

const DIMINISHERS = new Set([
  "slightly",
  "somewhat",
  "a_bit",
  "barely",
  "hardly",
  "sort_of",
  "kind_of"
]);

const MODALITY = {
  must: "obligation",
  should: "recommendation",
  ought: "recommendation",
  may: "permission",
  might: "possibility",
  could: "possibility",
  can: "ability",
  will: "future",
  would: "conditional"
};

const TEMPORAL_CUES = {
  yesterday: "past",
  today: "present",
  tomorrow: "future",
  now: "present",
  currently: "present",
  soon: "future",
  later: "future",
  previously: "past",
  recently: "past"
};

const PRONOUNS = new Set([
  "i",
  "you",
  "he",
  "she",
  "it",
  "we",
  "they",
  "me",
  "him",
  "her",
  "us",
  "them",
  "my",
  "your",
  "his",
  "their",
  "our"
]);

const DETERMINERS = new Set([
  "a",
  "an",
  "the",
  "this",
  "that",
  "these",
  "those",
  "each",
  "every",
  "some",
  "any",
  "no",
  "all",
  "most"
]);

const QUANTIFIERS = {
  all: "universal",
  every: "universal",
  each: "distributive",
  many: "plural",
  few: "small_amount",
  some: "existential",
  any: "open",
  several: "plural",
  most: "majority",
  none: "zero"
};

const QUESTION_WORDS = new Set(["who", "what", "when", "where", "why", "how", "which"]);

const SENTIMENT_POSITIVE = new Set([
  "happy",
  "joy",
  "love",
  "great",
  "excellent",
  "good",
  "wonderful",
  "amazing",
  "success",
  "positive"
]);

const SENTIMENT_NEGATIVE = new Set([
  "sad",
  "angry",
  "hate",
  "bad",
  "terrible",
  "awful",
  "horrible",
  "failure",
  "negative",
  "poor"
]);

const COMMON_VERBS = new Set([
  "be",
  "am",
  "is",
  "are",
  "was",
  "were",
  "have",
  "has",
  "had",
  "do",
  "does",
  "did",
  "say",
  "make",
  "go",
  "get",
  "see",
  "know",
  "think",
  "take",
  "come",
  "give",
  "find",
  "tell",
  "work",
  "call",
  "try",
  "ask",
  "need",
  "feel",
  "become",
  "leave",
  "put",
  "mean",
  "keep",
  "let",
  "begin",
  "help",
  "talk",
  "turn",
  "like",
  "start",
  "show",
  "hear",
  "play",
  "run",
  "move",
  "live",
  "believe",
  "bring",
  "happen",
  "write",
  "provide",
  "sit",
  "stand",
  "lose",
  "pay",
  "meet",
  "include",
  "continue",
  "set",
  "learn",
  "change",
  "lead",
  "understand",
  "watch",
  "follow",
  "stop",
  "create",
  "speak",
  "read",
  "allow",
  "add",
  "spend",
  "grow",
  "open",
  "walk",
  "win",
  "teach"
]);

const ADJECTIVE_SUFFIXES = ["ous", "ful", "able", "ible", "al", "ive", "less", "ic", "ish"];
const ADVERB_SUFFIXES = ["ly"];
const VERB_SUFFIXES = ["ed", "ing", "en", "ify", "ise", "ize"];

const WORD_RE = /[A-Za-z']+|[0-9]+|[^\sA-Za-z0-9]/g;

const tokenize = (sentence) => {
  const matches = sentence.match(WORD_RE) ?? [];
  return matches.map((match, index) => ({
    text: match,
    lemma: match.toLowerCase().replace(/'/g, ""),
    index
  }));
};

const isContentWord = (lemma) =>
  /^[a-z]+$/.test(lemma) && !DETERMINERS.has(lemma) && !PRONOUNS.has(lemma);

const guessPos = (lemma) => {
  if (COMMON_VERBS.has(lemma)) return "verb";
  if (lemma.endsWith("s") && COMMON_VERBS.has(lemma.slice(0, -1))) return "verb";
  if (VERB_SUFFIXES.some((suffix) => lemma.endsWith(suffix))) return "verb";
  if (ADVERB_SUFFIXES.some((suffix) => lemma.endsWith(suffix))) return "adverb";
  if (ADJECTIVE_SUFFIXES.some((suffix) => lemma.endsWith(suffix))) return "adjective";
  if (DETERMINERS.has(lemma)) return "determiner";
  if (PRONOUNS.has(lemma)) return "pronoun";
  if (QUESTION_WORDS.has(lemma)) return "question_word";
  if (/^\d+$/.test(lemma)) return "number";
  if (!/^[a-z]+$/.test(lemma)) return "punctuation";
  return "noun";
};

const detectSentenceType = (tokens) => {
  if (tokens.some((token) => token.text === "?")) return "interrogative";
  if (tokens.some((token) => token.text === "!")) return "exclamatory";
  return "declarative";
};

const estimateTense = (tokens) => {
  if (tokens.some((token) => token.text.toLowerCase().endsWith("ed"))) return "past";
  if (tokens.some((token) => token.lemma === "will")) return "future";
  if (tokens.some((token) => ["was", "were", "had"].includes(token.lemma))) return "past";
  return "present";
};

const pickMainVerb = (tokens) => tokens.find((token) => guessPos(token.lemma) === "verb") ?? null;

const guessSubject = (tokens, verbIndex) => {
  if (verbIndex === null) return null;
  const candidates = tokens
    .slice(0, verbIndex)
    .filter((token) => ["noun", "pronoun"].includes(guessPos(token.lemma)));
  if (!candidates.length) return null;
  return candidates.slice(-2).map((token) => token.text).join(" ");
};

const guessObject = (tokens, verbIndex) => {
  if (verbIndex === null) return null;
  const candidates = tokens
    .slice(verbIndex + 1)
    .filter((token) => ["noun", "pronoun"].includes(guessPos(token.lemma)));
  if (!candidates.length) return null;
  return candidates.slice(0, 2).map((token) => token.text).join(" ");
};

const sentimentScore = (tokens) =>
  tokens.reduce((score, token) => {
    if (SENTIMENT_POSITIVE.has(token.lemma)) return score + 1;
    if (SENTIMENT_NEGATIVE.has(token.lemma)) return score - 1;
    return score;
  }, 0);

const sentimentLabel = (score) => {
  if (score > 0) return "positive";
  if (score < 0) return "negative";
  return "neutral";
};

const collectTopics = (tokens) => {
  const topics = [];
  tokens.forEach((token) => {
    if (isContentWord(token.lemma) && /^[a-z]+$/.test(token.lemma)) {
      topics.push(token.lemma);
    }
  });
  return Array.from(new Set(topics));
};

const detectModality = (tokens) =>
  tokens.flatMap((token) => (MODALITY[token.lemma] ? [MODALITY[token.lemma]] : []));

const detectPolarity = (tokens) =>
  tokens.some((token) => NEGATIONS.has(token.lemma)) ? "negative" : "affirmative";

const temporalCues = (tokens) =>
  tokens.flatMap((token) => (TEMPORAL_CUES[token.lemma] ? [token.lemma] : []));

const quantifiers = (tokens) =>
  tokens.flatMap((token) => (QUANTIFIERS[token.lemma] ? [token.lemma] : []));

const wordInfluence = (tokens) =>
  tokens.map((token) => {
    const lemma = token.lemma;
    const pos = guessPos(lemma);
    let influence = "background";
    let weight = 0.2;

    if (isContentWord(lemma)) {
      influence = "content";
      weight = 1.0;
    }
    if (NEGATIONS.has(lemma)) {
      influence = "negation";
      weight = 1.4;
    }
    if (INTENSIFIERS.has(lemma)) {
      influence = "intensifier";
      weight = 1.2;
    }
    if (DIMINISHERS.has(lemma)) {
      influence = "diminisher";
      weight = 0.6;
    }
    if (QUANTIFIERS[lemma]) {
      influence = "quantifier";
      weight = 1.1;
    }
    if (MODALITY[lemma]) {
      influence = "modality";
      weight = 1.1;
    }

    return {
      token: token.text,
      lemma,
      index: token.index,
      pos_guess: pos,
      influence_type: influence,
      weight
    };
  });

export const analyzeSentence = (sentence) => {
  const tokens = tokenize(sentence);
  const mainVerb = pickMainVerb(tokens);
  const verbIndex = mainVerb ? mainVerb.index : null;
  const score = sentimentScore(tokens);

  return {
    sentence,
    sentence_type: detectSentenceType(tokens),
    polarity: detectPolarity(tokens),
    modality: detectModality(tokens),
    tense: estimateTense(tokens),
    temporal_cues: temporalCues(tokens),
    quantifiers: quantifiers(tokens),
    main_predicate: mainVerb ? mainVerb.text : null,
    subject: guessSubject(tokens, verbIndex),
    object: guessObject(tokens, verbIndex),
    topics: collectTopics(tokens),
    sentiment: {
      score,
      label: sentimentLabel(score)
    },
    word_influence: wordInfluence(tokens)
  };
};
